### PR TITLE
Contact duration and distance

### LINKED ISSUE #122: Distance and duration handling
- Modify DB TCN records to `{ TCN, start_timestamp, min_distance, avg_distance, end_timestamp }`
- Aggregate TCNs during matching, considering contiguity.
- Add duration and distance fields to alerts.

To be determined:

- To pass TCNs to core, apps can either pass individual TCNs (with distance) or batches. Individual TCNs would be ideal, as there would be no duplicated logic in the apps, but we've to confirm FFI/JNI performance.
- Confirm whether we need min / avg distance or only min.

More [context](https://github.com/Co-Epi/app-ios/issues/158)

### PR DESCRIPTION
**Pull request by ivnsch**

### CODE REVIEW

File: src/reports_updater.rs [hunk 1]
```diff
@@ -293,27 +378,76 @@ where
...
+                Some(last_group) => {
+                    if tcn.contact_start.value as i64 - last_contact_end < separator as i64 {
+                        last_group.push(tcn)
+                    } else {
+                        exposures.push(Exposure::create(tcn));
+                    }
+                }
+                None => exposures.push(Exposure::create(tcn)),
+            }
+            // if tcn.contact_start.value as i64 - last_contact_end < separator as i64 {
```
Comment (duskoo) [inline src/reports_updater.rs:440]: remove

File: src/reports_updater.rs [hunk 2]
```diff
@@ -293,27 +378,76 @@ where
...
+            id: format!("{:?}", signed_report.sig),
             report: public_report,
-            contact_time: matched_report.contact_time.value,
+            contact_start: measurements.contact_start,
+            contact_end: measurements.contact_end,
+            min_distance: measurements.min_distance,
         })
     }
 
+    fn group_in_exposures(mut tcns: Vec<ObservedTcn>, separator: u64) -> Vec<Exposure> {
```
Comment (duskoo) [inline src/reports_updater.rs]: Naming: I prefer threshold or tolerance to separator. Separator is usually a character used to split a string.

**Comment by duskoo** [comment]
@i-schuetz Ups, I noticed its a draft PR only after I have gone through the whole thing :(
Looks good by the way 👍

File: src/reports_updater.rs (hunk 2 above)
Comment (ivnsch) [reply src/reports_updater.rs]: I agree, threshold "magnitude or intensity that must be _exceeded_" is more accurate.

File: src/reports_updater.rs [hunk 3]
```diff
@@ -180,22 +350,53 @@ impl TcnDaoImpl {
...
             .map_err(ServicesError::from)
     }
 
-    fn save(&self, observed_tcn: &ObservedTcn) -> Result<(), ServicesError> {
-        let tcn_str = hex::encode(observed_tcn.tcn.0);
+    fn find_tcns(
+        &self,
+        _with: Vec<TemporaryContactNumber>,
+    ) -> Result<Vec<ObservedTcn>, ServicesError> {
+        // TODO: IN query
```
Comment (i-schuetz) [inline src/reports_updater.rs:365]: I'll do this in a separate PR. `self.all` is a valid substitute, as the TCNs are filtered programmatically, so this is just a performance optimization, and it's more important to start testing with the apps.

File: src/ios/c_headers/coepicore.h [hunk 4]
```diff
@@ -3,96 +3,150 @@
...
 
+#if (defined(TARGET_OS_IOS) || defined(TARGET_OS_MACOS))
 CFStringRef submit_symptoms(void);
+#endif
 
+#if (defined(TARGET_OS_IOS) || defined(TARGET_OS_MACOS))
 int32_t trigger_callback(const char *my_str);
+#endif
 
+#if (defined(TARGET_OS_IOS) || defined(TARGET_OS_MACOS))
```
Comment (duskoo) [inline src/ios/c_headers/coepicore.h:150]: It would be nice to be able to distinguish methods used for testing/validation (ie. add prefix test_).

Comment (ivnsch) [reply src/ios/c_headers/coepicore.h:150]: Open an issue?

**Comment by i-schuetz** [comment]
Sorry, I forgot the TCNs merging when adding them to the in-memory batch. Converting back to draft.

File: src/reports_updater.rs [hunk 5]
```diff
@@ -96,45 +104,242 @@ impl TcnMatcherRayon {
...
+    pub fn new(tcn_dao: Arc<T>, exposure_grouper: ExposureGrouper) -> TcnBatchesManager<T> {
+        TcnBatchesManager {
+            tcn_dao,
+            tcns_batch: Mutex::new(HashMap::new()),
+            exposure_grouper,
+        }
+    }
+
+    pub fn flush(&self) -> Result<(), ServicesError> {
+        let res = self.tcns_batch.lock();
```
Comment (ivnsch) [inline src/reports_updater.rs]: I don't like this. `flush` is executed in a background thread, but the lock will block anyway the threads pushing the TCNs. The later seem to belong to BLE, at least on iOS.

Probably it makes sense to clear and release the lock immediately, and execute the db update after.

File: src/ios/c_headers/coepicore.h (hunk 4 above)
Comment (duskoo) [reply src/ios/c_headers/coepicore.h:150]: #126

File: src/reports_updater.rs (hunk 5 above)
Comment (duskoo) [reply src/reports_updater.rs]: @i-schuetz  That could improve performance a bit. New ticket?

File: src/reports_updater.rs [hunk 6]
```diff
@@ -96,45 +104,242 @@ impl TcnMatcherRayon {
...
+        instance
+    }
+
+    fn schedule_process_batches(tcn_batches_manager: Arc<Mutex<TcnBatchesManager<T>>>) -> TimerData {
+        let timer = Arc::new(Mutex::new(Timer::new()));
+        TimerData {
+            _timer: timer.clone(),
+            _guard: timer.clone().lock().unwrap().schedule_repeating(chrono::Duration::seconds(10), move || {
+                debug!("Flushing TCN batches into database");
+                let tcn_batches_manager_res = tcn_batches_manager.lock();
```
Comment (duskoo) [inline src/reports_updater.rs]: Another lock

File: src/reports_updater.rs [hunk 7]
```diff
@@ -96,45 +104,242 @@ impl TcnMatcherRayon {
...
+            // Matching exposures in DB
+            Some(db_tcns) => {
+                if let Some(last) = db_tcns.last() {
+
+                    // If contiguous to last DB exposure, merge with it, otherwise append.
+                    let tail = match Self::merge_tcns(&exposure_grouper, last.to_owned(), tcn.clone()) {
+                        Some(merged ) => vec![merged],
+                        None => vec![last.to_owned(), tcn]
+                    };
+                    let mut head: Vec<ObservedTcn> = db_tcns.to_owned().into_iter().take(db_tcns.len() - 1).collect();
```
Comment (duskoo) [inline src/reports_updater.rs:223]: We could avoid iterating over the vector and just replace the last element: https://doc.rust-lang.org/std/mem/fn.replace.html

**Review by duskoo — APPROVED** [review]

File: src/reports_updater.rs (hunk 6 above)
Comment (ivnsch) [reply src/reports_updater.rs]: This one can be removed 👍

File: src/reports_updater.rs (hunk 7 above)
Comment (ivnsch) [reply src/reports_updater.rs:223]: Unclear how `mem::replace` helps. It seems to replace the complete vector. `db_tcns` is also unlikely to contain more than a couple of elements (these are the stored, averaged exposures for an individual TCN) so the performance is negligible.

File: src/reports_updater.rs (hunk 5 above)
Comment (ivnsch) [reply src/reports_updater.rs]: Done: https://github.com/Co-Epi/app-backend-rust/pull/123/commits/f2304da8059d0f691ad3ebf6f67095042af264f8 / https://github.com/Co-Epi/app-backend-rust/pull/123/commits/82f19dd4939d380884f69879b03d28c2330dc753

File: src/reports_updater.rs (hunk 7 above)
Comment (duskoo) [reply src/reports_updater.rs:223]: mem::replace works on a vector element. 
Good point regarding small vector size.

**Pull request merged**